File: Java/src/com/loggersoft/performance/Position.java
package com.loggersoft.performance;

public class Position {
    private double lat;
    private double lon;

    public Position(double lat, double lon) {
        this.lat = lat;
        this.lon = lon;
    }

    public boolean isValid() {
        return (lat >= -90.0 && lat <= 90.0 && lon >= -180.0 && lon <= 180.0);
    }

    public double getLat() {
        return lat;
    }

    public void setLat(double lat) {
        this.lat = lat;
    }

    public double getLon() {
        return lon;
    }

    public void setLon(double lon) {
        this.lon = lon;
    }

    public double getDistance(Position other) {
        final double lat1 = Math.toRadians(lat);
        final double lon1 = Math.toRadians(lon);
        final double lat2 = Math.toRadians(other.getLat());
        final double lon2 = Math.toRadians(other.getLon());
        final double sinLat = Math.sin((lat2 - lat1) / 2.0);
        final double sinLon = Math.sin((lon2 - lon1) / 2.0);
        final double a = sinLat * sinLat + Math.cos(lat1) * Math.cos(lat2) * sinLon * sinLon;
        return 2.0 * Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a));
    }
}

File: Scala/src/main/scala/com/loggersoft/scala/performance/Position.scala
package com.loggersoft.scala.performance

import math.toRadians
import math.sin
import math.cos
import math.sqrt
import math.atan2

class Position(val lat: Double, val lon: Double) {

  def isValid()
    = lat >= -90.0 && lat <= 90 && lon >= -180 && lon <= 180

  def getDistance(other: Position): Double = {
    val lat1 = toRadians(lat)
    val lon1 = toRadians(lon)
    val lat2 = toRadians(other.lat)
    val lon2 = toRadians(other.lon)
    val sinLat = sin((lat2 - lat1) / 2.0)
    val sinLon = sin((lon2 - lon1) / 2.0)
    val a = sinLat * sinLat + cos(lat1) * cos(lat2) * sinLon * sinLon
    return 2.0 * atan2(sqrt(a), sqrt(1.0 - a))
  }
}
